kaiax/reward/impl/getter.go

```diff
@@ -0,0 +1,395 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Kaia library. If not, see <http://www.gnu.org/licenses/>.
+
+package impl
+
+import (
+	"math/big"
+
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/common"
+	"github.com/kaiachain/kaia/common/math"
+	"github.com/kaiachain/kaia/kaiax/reward"
+	"github.com/kaiachain/kaia/kaiax/staking"
+)
+
+// GetRewardSummary retrospectively calculates the reward summary at the given block number.
+func (r *RewardModule) GetRewardSummary(num uint64) (*reward.RewardSummary, error) {
+	config, _, totalFee, err := r.loadBlockData(num)
+	if err != nil {
+		return nil, err
+	}
+	return getRewardSummary(config, totalFee), nil
+}
+
+func (r *RewardModule) loadBlockData(num uint64) (*reward.RewardConfig, *types.Header, *big.Int, error) {
+	block := r.Chain.GetBlockByNumber(num)
+	if block == nil {
+		return nil, nil, nil, reward.ErrNoBlock
+	}
+	receipts := r.Chain.GetReceiptsByBlockHash(block.Hash())
+	if receipts == nil {
+		return nil, nil, nil, reward.ErrNoReceipts
+	}
+	header := block.Header()
+	txs := block.Transactions()
+
+	config, err := reward.NewRewardConfig(r.ChainConfig, r.GovModule, header)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	totalFee, err := getTotalFee(config, header, txs, receipts)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return config, header, totalFee, nil
+}
+
+func getRewardSummary(config *reward.RewardConfig, totalFee *big.Int) *reward.RewardSummary {
+	minted := new(big.Int).Set(config.MintingAmount)
+
+	burntFee := big.NewInt(0)
+	if config.IsSimple { // simplified getDeferredRewardSimple
+		if config.Rules.IsMagma {
+			burntFee = getBurnAmountMagma(totalFee)
+		}
+	} else { // simplified getDeferredRewardFull
+		if config.Rules.IsKore {
+			burntFee = getBurnAmountKore(config, totalFee)
+		} else if config.Rules.IsMagma {
+			burntFee = getBurnAmountMagma(totalFee)
+		}
+	}
+
+	return &reward.RewardSummary{
+		Minted:   minted,
+		TotalFee: totalFee,
+		BurntFee: burntFee,
+	}
+}
+
+// GetBlockReward retrospectively calculates the block reward distributed at the given block number.
+// The result includes both non-deferred and deferred fees, so it may differ from GetDeferredReward.
+func (r *RewardModule) GetBlockReward(num uint64) (*reward.RewardSpec, error) {
+	config, header, totalFee, err := r.loadBlockData(num)
+	if err != nil {
+		return nil, err
+	}
+
+	spec, err := r.getDeferredReward(config, header, totalFee)
+	if err != nil {
+		return nil, err
+	}
+
+	return r.specWithNonDeferredFee(spec, config, header, totalFee)
+}
+
+// specWithNonDeferredFee adds non-deferred fees to the reward spec.
+func (r *RewardModule) specWithNonDeferredFee(spec *reward.RewardSpec, config *reward.RewardConfig, header *types.Header, totalFee *big.Int) (*reward.RewardSpec, error) {
```

Comment:
Can we use a copy of the parameter `spec` so it does not change? Like big int operations (Add, etc.) where parameters don't change.
It seems strange to both change the parameter spec and return the same pointer.

Reply:
Fixed in deb4cb8.